# PharmaOps-AI
## Data Extraction & Understanding

### Objective
The objective of this notebook is to connect to the MySQL database, extract data from all required tables, and perform an initial understanding of the dataset before preprocessing and analysis.

### Dataset
- Medicines_Master
- Medicines_Inventory
- Suppliers_Master
- Sales_Transactions
- Waste_Records
- Category_Master

In [2]:
!pip install mysql-connector-python sqlalchemy

Defaulting to user installation because normal site-packages is not writeable
   ---------------------------------------- 0.0/18.2 MB ? eta -:--:--
   ---- ----------------------------------- 1.8/18.2 MB 23.4 MB/s eta 0:00:01
   -------------- ------------------------- 6.6/18.2 MB 23.8 MB/s eta 0:00:01
   ------------------------- -------------- 11.5/18.2 MB 23.3 MB/s eta 0:00:01
   --------------------------------- ------ 15.2/18.2 MB 21.7 MB/s eta 0:00:01
   ---------------------------------------  18.1/18.2 MB 21.7 MB/s eta 0:00:01
   ---------------------------------------- 18.2/18.2 MB 19.4 MB/s  0:00:01
   ---------------------------------------- 0.0/2.1 MB ? eta -:--:--
   ---------------------------------------- 2.1/2.1 MB 19.7 MB/s  0:00:00

   ---------------------------------------- 0/3 [mysql-connector-python]
   ---------------------------------------- 0/3 [mysql-connector-python]
   ---------------------------------------- 0/3 [mysql-connector-python]
   -----------------

In [5]:
import sys
print(sys.executable)

C:\Users\venka\anaconda3\envs\myenv\python.exe


In [6]:
import sys
print(sys.version)

3.10.19 | packaged by Anaconda, Inc. | (main, Oct 21 2025, 16:41:31) [MSC v.1929 64 bit (AMD64)]


In [7]:
!{sys.executable} -m pip list

Package           Version
----------------- -----------
asttokens         3.0.1
attrs             26.1.0
certifi           2026.5.20
cffi              2.0.0
colorama          0.4.6
comm              0.2.3
contourpy         1.3.2
cycler            0.12.1
debugpy           1.8.20
decorator         5.2.1
exceptiongroup    1.3.1
executing         2.2.1
fonttools         4.63.0
h11               0.16.0
idna              3.18
ipykernel         7.2.0
ipython           8.38.0
jedi              0.19.2
jupyter_client    8.8.0
jupyter_core      5.9.1
kiwisolver        1.5.0
matplotlib        3.10.9
matplotlib-inline 0.2.1
nest-asyncio      1.6.0
numpy             2.2.6
outcome           1.3.0.post0
packaging         25.0
pandas            2.3.3
parso             0.8.6
pillow            12.3.0
pip               26.1.2
platformdirs      4.9.2
prompt_toolkit    3.0.52
psutil            7.2.2
pure_eval         0.2.3
pycparser         3.0
Pygments          2.19.2
pyparsing         3.3.2
PySocks       

In [8]:
import sys

!{sys.executable} -m pip install mysql-connector-python sqlalchemy

   ---------------------------------------- 0.0/17.7 MB ? eta -:--:--
   ---------------------------------------- 0.0/17.7 MB ? eta -:--:--
    --------------------------------------- 0.3/17.7 MB ? eta -:--:--
   - -------------------------------------- 0.5/17.7 MB 1.9 MB/s eta 0:00:10
   - -------------------------------------- 0.8/17.7 MB 1.9 MB/s eta 0:00:10
   -- ------------------------------------- 1.0/17.7 MB 1.4 MB/s eta 0:00:13
   -- ------------------------------------- 1.3/17.7 MB 1.2 MB/s eta 0:00:14
   --- ------------------------------------ 1.6/17.7 MB 1.3 MB/s eta 0:00:13
   --- ------------------------------------ 1.6/17.7 MB 1.3 MB/s eta 0:00:13
   ---- ----------------------------------- 1.8/17.7 MB 1.2 MB/s eta 0:00:14
   ---- ----------------------------------- 2.1/17.7 MB 1.2 MB/s eta 0:00:14
   ---- ----------------------------------- 2.1/17.7 MB 1.2 MB/s eta 0:00:14
   ----- ---------------------------------- 2.4/17.7 MB 1.1 MB/s eta 0:00:15
   ----- -----------


[notice] A new release of pip is available: 26.1.2 -> 26.2
[notice] To update, run: C:\Users\venka\anaconda3\envs\myenv\python.exe -m pip install --upgrade pip


In [11]:
import sys
print(sys.executable)

!{sys.executable} -m pip install mysql-connector-python sqlalchemy -v

C:\Users\venka\anaconda3\envs\myenv\python.exe
Using pip 26.1.2 from C:\Users\venka\anaconda3\envs\myenv\lib\site-packages\pip (python 3.10)



[notice] A new release of pip is available: 26.1.2 -> 26.2
[notice] To update, run: C:\Users\venka\anaconda3\envs\myenv\python.exe -m pip install --upgrade pip


In [1]:
import pandas as pd
import numpy as np

import mysql.connector
from sqlalchemy import create_engine

import warnings
warnings.filterwarnings("ignore")

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 20)

print("Libraries imported successfully.")

Libraries imported successfully.


In [3]:
HOST = "localhost"
PORT = 3306
USER = "root"
PASSWORD = "Venkat#26158"
DATABASE = "PharmaOps_AI"

engine = create_engine(
    f"mysql+mysqlconnector://{USER}:{PASSWORD}@{HOST}:{PORT}/{DATABASE}"
)

print("Database connection established successfully.")

Database connection established successfully.


## Step 2: Load Data Sources

The project uses data from two sources:

- MySQL Database
- Final Medicines Master CSV

The Medicines Master dataset is loaded from the finalized CSV file, while the remaining operational tables are loaded directly from MySQL.

In [15]:
# Load Medicines Master from CSV
medicines_df = pd.read_csv("../../data/processed/Medicines_Master_Final.csv")

# Load remaining tables from MySQL
inventory_df = pd.read_sql("SELECT * FROM Medicines_Inventory", engine)
suppliers_df = pd.read_sql("SELECT * FROM Suppliers_Master", engine)
sales_df = pd.read_sql("SELECT * FROM Sales_Transactions", engine)
waste_df = pd.read_sql("SELECT * FROM Waste_Records", engine)
category_df = pd.read_sql("SELECT * FROM Category_Master", engine)

print("All datasets loaded successfully.")

All datasets loaded successfully.


In [13]:
import pandas as pd

medicines_df = pd.read_csv("../../data/processed/Medicines_Master_Final.csv")

print(medicines_df.shape)
medicines_df.head()

(111815, 11)


,Medicine_ID,Product_NDC,Generic_Name,Brand_Name,Manufacturer,Active_Ingredient,Strength,Dosage_Form,Route,Pharm_Class,Category_ID
0,MED000001,72162-1557,Nabumetone,Nabumetone,Bryant Ranch Prepack,NABUMETONE,500 mg/1,TABLET,ORAL,Anti-Inflammatory Agents,CAT004
1,MED000002,72162-1561,Benzonatate,Benzonatate,Bryant Ranch Prepack,BENZONATATE,100 mg/1,CAPSULE,ORAL,Decreased Tracheobronchial Stretch Receptor Ac...,CAT006
2,MED000003,72162-1610,Rivastigmine,Rivastigmine Transdermal System,Bryant Ranch Prepack,RIVASTIGMINE,13.3 mg/24h,"PATCH, EXTENDED RELEASE",TRANSDERMAL,Cholinesterase Inhibitor [EPC],CAT005
3,MED000004,72162-1612,Amphetamine Sulfate,Amphetamine Sulfate,Bryant Ranch Prepack,AMPHETAMINE SULFATE,5 mg/1,TABLET,ORAL,Central Nervous System Stimulant [EPC],CAT005
4,MED000005,72162-1625,Spironolactone And Hydrochlorothiazide,Spironolactone And Hydrochlorothiazide,Bryant Ranch Prepack,HYDROCHLOROTHIAZIDE,25 mg/1,TABLET,ORAL,Aldosterone Antagonist [EPC],CAT003


# Step 5: Data Understanding

In this section, we explore the structure and quality of each dataset by examining:
- Dataset dimensions
- Column names
- Data types
- Missing values
- Duplicate records
- Summary statistics

This helps us understand the data before preprocessing and analysis.

In [17]:
datasets = {
    "Medicines_Master": medicines_df,
    "Medicines_Inventory": inventory_df,
    "Suppliers_Master": suppliers_df,
    "Sales_Transactions": sales_df,
    "Waste_Records": waste_df,
    "Category_Master": category_df
}

In [18]:
for name, df in datasets.items():

    print("=" * 70)
    print(f"{name}")
    print("=" * 70)

    print(f"Rows    : {df.shape[0]}")
    print(f"Columns : {df.shape[1]}")

    print("\nColumn Names:")
    print(df.columns.tolist())

    print("\n")

Medicines_Master
Rows    : 111815
Columns : 11

Column Names:
['Medicine_ID', 'Product_NDC', 'Generic_Name', 'Brand_Name', 'Manufacturer', 'Active_Ingredient', 'Strength', 'Dosage_Form', 'Route', 'Pharm_Class', 'Category_ID']


Medicines_Inventory
Rows    : 4954
Columns : 14

Column Names:
['Inventory_ID', 'Medicine_ID', 'Supplier_ID', 'Batch_Number', 'Manufacturing_Date', 'Expiry_Date', 'Quantity_In_Stock', 'Unit_Cost', 'Selling_Price', 'Reorder_Level', 'Storage_Temperature', 'Warehouse_Location', 'Stock_Status', 'Last_Restock_Date']


Suppliers_Master
Rows    : 75
Columns : 15

Column Names:
['Supplier_ID', 'Supplier_Name', 'Supplier_Type', 'Supplier_Category', 'Preferred_Supplier', 'City', 'State', 'Contact_Email', 'Contact_Number', 'Lead_Time_Days', 'Supplier_Rating', 'Active_Status', 'Contract_Start_Date', 'Contract_End_Date', 'GST_Number']


Sales_Transactions
Rows    : 9901
Columns : 15

Column Names:
['Transaction_ID', 'Invoice_Number', 'Medicine_ID', 'Branch_ID', 'Transaction_

## Data Types and General Information

This step provides information about each dataset including data types, non-null values, and memory usage.

In [19]:
for name, df in datasets.items():

    print("=" * 70)
    print(f"{name}")
    print("=" * 70)

    df.info()

    print("\n")

Medicines_Master
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 111815 entries, 0 to 111814
Data columns (total 11 columns):
 #   Column             Non-Null Count   Dtype 
---  ------             --------------   ----- 
 0   Medicine_ID        111815 non-null  object
 1   Product_NDC        111815 non-null  object
 2   Generic_Name       111815 non-null  object
 3   Brand_Name         111815 non-null  object
 4   Manufacturer       111815 non-null  object
 5   Active_Ingredient  111815 non-null  object
 6   Strength           111815 non-null  object
 7   Dosage_Form        111815 non-null  object
 8   Route              111815 non-null  object
 9   Pharm_Class        111815 non-null  object
 10  Category_ID        111815 non-null  object
dtypes: object(11)
memory usage: 9.4+ MB


Medicines_Inventory
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4954 entries, 0 to 4953
Data columns (total 14 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --

## Missing Value Analysis

Identify missing values in every dataset to understand data completeness.

In [20]:
for name, df in datasets.items():

    print("=" * 70)
    print(name)
    print("=" * 70)

    missing = df.isnull().sum()

    missing = missing[missing > 0]

    if missing.empty:
        print("No Missing Values\n")
    else:
        print(missing)

    print()

Medicines_Master
No Missing Values


Medicines_Inventory
No Missing Values


Suppliers_Master
No Missing Values


Sales_Transactions
No Missing Values


Waste_Records
No Missing Values


Category_Master
No Missing Values




## Duplicate Record Analysis

Check for duplicate rows in each dataset.

In [21]:
for name, df in datasets.items():

    duplicates = df.duplicated().sum()

    print(f"{name} : {duplicates} duplicate rows")

Medicines_Master : 0 duplicate rows
Medicines_Inventory : 0 duplicate rows
Suppliers_Master : 0 duplicate rows
Sales_Transactions : 0 duplicate rows
Waste_Records : 0 duplicate rows
Category_Master : 0 duplicate rows


## Summary Statistics

Generate descriptive statistics for numerical columns to understand distributions and identify potential anomalies.

In [22]:
for name, df in datasets.items():

    print("=" * 70)
    print(name)
    print("=" * 70)

    display(df.describe())

Medicines_Master


,Medicine_ID,Product_NDC,Generic_Name,Brand_Name,Manufacturer,Active_Ingredient,Strength,Dosage_Form,Route,Pharm_Class,Category_ID
count,111815,111815,111815,111815,111815,111815,111815,111815,111815,111815,111815
unique,111815,111815,15147,39104,7148,3951,5461,137,64,535,14
top,MED137042,76354-167,Zinc Oxide,Ibuprofen,"Hahnemann Laboratories, Inc.",ACETAMINOPHEN,10 mg/1,TABLET,ORAL,UNKNOWN,CAT015
freq,1,1,2006,648,4807,3495,3970,18015,65750,41004,41004


Medicines_Inventory


,Quantity_In_Stock,Unit_Cost,Selling_Price,Reorder_Level
count,4954.000000,4954.000000,4954.000000,4954.000000
mean,499.672184,299.438020,389.006952,112.714170
std,284.629572,172.542195,225.003269,50.189528
min,0.000000,5.120000,6.180000,25.000000
25%,261.000000,152.287500,196.325000,70.000000
50%,498.000000,295.085000,380.890000,113.000000
75%,748.000000,448.425000,580.515000,155.000000
max,1000.000000,599.990000,835.390000,200.000000


Suppliers_Master


,Lead_Time_Days,Supplier_Rating
count,75.000000,75.000000
mean,8.453333,4.349333
std,4.754211,0.373578
min,2.000000,3.500000
25%,5.000000,4.150000
50%,7.000000,4.400000
75%,10.500000,4.600000
max,20.000000,5.000000


Sales_Transactions


,Sale_Time,Quantity_Sold,Unit_Selling_Price,Discount_Percentage,Total_Amount
count,9901,9901.000000,9901.000000,9901.000000,9901.000000
mean,0 days 15:01:10.501969498,2.527219,606.367744,5.181800,1450.116119
std,0 days 04:04:37.646252573,1.941100,341.432962,6.006081,1519.194238
min,0 days 08:00:00,1.000000,10.010000,0.000000,9.120000
25%,0 days 11:27:00,1.000000,309.190000,0.000000,487.910000
50%,0 days 15:06:00,2.000000,610.870000,5.000000,939.250000
75%,0 days 18:33:00,3.000000,899.800000,10.000000,1901.800000
max,0 days 22:00:00,10.000000,1199.990000,20.000000,11560.300000


Waste_Records


,Quantity_Wasted,Unit_Cost,Total_Waste_Value
count,2481.000000,2481.000000,2481.000000
mean,91.881096,302.406485,27704.350512
std,58.606348,172.013302,25846.932648
min,1.000000,5.330000,5.600000
25%,39.000000,154.970000,6662.700000
50%,89.000000,300.140000,19652.180000
75%,142.000000,448.190000,41887.560000
max,200.000000,599.990000,119030.000000


Category_Master


,Category_ID,Category_Name,Description
count,15,15,15
unique,15,15,15
top,CAT001,Antibiotics,Medicines used to treat bacterial infections
freq,1,1,1


## Duplicate Record Analysis

Check for duplicate rows in each dataset to ensure data integrity before preprocessing.

In [23]:
for name, df in datasets.items():

    duplicates = df.duplicated().sum()

    print(f"{name:<25} : {duplicates}")

Medicines_Master          : 0
Medicines_Inventory       : 0
Suppliers_Master          : 0
Sales_Transactions        : 0
Waste_Records             : 0
Category_Master           : 0


## Initial Observations

### Key Findings

- Successfully loaded six datasets from both MySQL and the finalized Medicines Master CSV.
- The Medicines Master dataset contains **111,815 medicine records**, making it the largest dataset in the project.
- Inventory contains **4,954 records**, Sales contains **9,901 transactions**, Waste contains **2,481 records**, Suppliers contains **75 suppliers**, and Category Master contains **15 categories**.
- No missing values were identified in any dataset, indicating excellent data completeness.
- Data types are generally appropriate; however, several date columns are currently stored as object type and will be converted to datetime during preprocessing.
- Numerical fields such as inventory quantity, selling price, waste value, and sales amount are available for business analysis.
- Category Master acts as a lookup table containing 15 medicine categories.
- The datasets appear well-structured and are ready for preprocessing and feature engineering.